surface/include/pcl/surface/ball_pivoting.h

```diff
@@ -0,0 +1,230 @@
+     * find the center of newly pivoted ball, if empty, then pivoting is not successful
+     * @param is_back_first whether to consider assumed center on back surface first. it changes the result if
+     *        flipping between front surface and back surface is enabled (setAllowFlip)
+     * @param index index of triangle vertices, which are all one ball surface
+     * @param is_back_ball whether the pivoted ball is on back surface
+     * @return
+     */
+    boost::shared_ptr<Eigen::Vector3f>
+    getBallCenter (const bool is_back_first, std::vector<uint32_t> &index, bool &is_back_ball) const;
+  
+  public:
+    BallPivoting ();
+  
+    ~BallPivoting ();
+  
+    /**
+     * set the radius of ball
+     * @param radius
+     */
+    void
+    setSearchRadius (const double radius)
+    { radius_ = radius; }
+  
+    /**
+     * get the radius of ball
+     * @return
+     */
+    double
+    getSearchRadius () const
+    { return radius_; }
+  
+    /**
+     * set whether ball on the back surface is allowed. if this value is set to true, the ball only pivots on the
+     * direction of the normal vectors
+     * @param is_allow_back_ball
+     */
+    void
+    setAllowBackBall (const bool is_allow_back_ball)
+    { is_allow_back_ball_ = is_allow_back_ball; }
+  
+    /**
+     * get whether ball on the back surface is allowed
+     * @return
+     */
+    bool
+    getAllowBackBall () const
+    { return is_allow_back_ball_; }
+  
+    /**
+     * set whether flipping between front surface and back surface is allow. if not, the ball would only pivot
+     * so it stays on one side of the surface, depending on where the center of seed ball was found
+     * @param is_allow_flip
+     */
+    void
+    setAllowFlip (const bool is_allow_flip)
+    { is_allow_flip_ = is_allow_flip; }
+  
+    /**
+     * get whether flipping between front surface and back surface is allow
+     * @return
+     */
+    bool
+    getAllowFlip () const
+    { return is_allow_flip_; }
+  
+    /**
+     * Create the surface.
+     * @param output the resultant polygonal mesh
+     */
+    void
+    performReconstruction (pcl::PolygonMesh &output);
+  
+    /**
+     * Create the surface.
+     * @param points the vertex positions of the resulting mesh
+     * @param polygons the connectivity of the resulting mesh
+     */
+    virtual void
+    performReconstruction (pcl::PointCloud<PointNT> &points,
+                           std::vector<pcl::Vertices> &polygons);
+
+    /**
+     * estimate one radius and set it as ball radius. num_sample_point points would be selected randomly in cloud,
+     * as least ratio_success of the sample points must have at least num_point_in_radius neighbors within
+     * the estimated radius.
+     * @param num_sample_point
+     * @param num_point_in_radius
+     * @param ratio_success
+     */
+    virtual void
+    setEstimatedRadius (const int num_sample_point, const int num_point_in_radius, const float ratio_success);
```

Comment:
I think this is not a clear name. What about `setSearchRadiusAutomatic()`? This will make connection with the `setSearchRadius()` function, thus making clear for the user that these two functions have the same purpose of setting the search radius. Further, I think we should provide some reasonable default values for the arguments.